File: 4.1_stall4/stall4.cpp
/*
ID: dermotl1
LANG: C++
PROG: stall4
*/

#include <algorithm>
#include <cstring>
#include <fstream>
#include <iostream>
#include <limits>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

int cow_likes[201][201];
int cow_likes_count[201];
int final_match[201];
bool visited[201];
int n, m;
int result;

bool solve(int cow) {
  for (int i = 0; i < cow_likes_count[cow]; ++i) {
    int stall = cow_likes[cow][i];

    if (visited[stall] == false) {
      visited[stall] = true;
      if (final_match[stall] == 0 ||
          solve(final_match[stall])) {  // if stall isn't taken/other cow can be
                                        // reshuffled
        final_match[stall] = cow;
        return true;
      }
    }
  }
  return false;
}

int main() {
  ifstream fin("stall4.in");
  ofstream fout("stall4.out");

  fin >> n >> m;

  for (int i = 1; i <= n; ++i) {
    fin >> cow_likes_count[i];
    for (int j = 0; j < cow_likes_count[i]; ++j) {
      fin >> cow_likes[i][j];
    }
  }

  for (int i = 1; i <= n; ++i) {
    memset(visited, false, sizeof(visited));
    if (solve(i)) result++;
  }

  fout << result << "\n";

  return 0;
}

File: 4.2_stall4/stall4.java
/*
ID: dermotl1
LANG: JAVA
PROG: stall4
*/

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.*;

public class stall4 {
    static int cow_likes[][] = new int[201][201];
    static int cow_likes_count[] = new int[201];
    static int final_match[] = new int[201];
    static Boolean visited[] = new Boolean[201];
    static int n, m;
    static int result;

    public static boolean solve(int cow) {
        for (int i = 0; i < cow_likes_count[cow]; ++i) {
            int stall = cow_likes[cow][i];

            if (!visited[stall]) {
                visited[stall] = true;
                if (final_match[stall] == 0 || solve(final_match[stall])) {
                    final_match[stall] = cow;
                    return true;
                }
            }
        }
        return false;
    }
    public static void main(String[] args) throws FileNotFoundException, IOException {
        Scanner in = new Scanner(new File("stall4.in"));
        PrintWriter out = new PrintWriter(new File("stall4.out"));

        n = in.nextInt();
        m = in.nextInt();

        for (int i = 1; i <= n; ++i) {
            cow_likes_count[i] = in.nextInt();
            for (int j = 0; j < cow_likes_count[i]; ++j) {
                cow_likes[i][j] = in.nextInt();
            }
        }

        for (int i = 1; i <= n; ++i) {
            Arrays.fill(visited, false);
            if (solve(i)) result++;
        }

        out.println(result);

        in.close();
        out.close();
    }
}
